### ISSUE TITLE
Improve performance of Elastic query to get facet values

### ISSUE THREAD

**Issue by JeroenDeDauw** [ProfessionalWiki/WikibaseFacetedSearch#27]
Follow up to
* https://github.com/ProfessionalWiki/WikibaseFacetedSearch/issues/23

From https://chatgpt.com/share/67531a04-54ec-800d-86a6-b01a3cd4f11d:

> When querying a Wikibase instance using Elasticsearch and WikibaseCirrusSearch, we need efficient aggregations on the statement_keywords field to extract and count specific property values (e.g., top authors, volumes). However, statement_keywords is a text field, which doesn't support aggregations by default. The challenge is to perform these aggregations efficiently, especially with large datasets.

1. **Enable Fielddata on statement_keywords**: Modify the Elasticsearch index mapping to set `fielddata: true` on the statement_keywords field, allowing aggregations on this text field.
2. **Use Runtime Fields in Queries**: Define runtime fields within the query to extract desired property values at query time, avoiding changes to the index mapping.
3. **Reindex Data to Include Aggregatable Fields**: Modify data ingestion to extract property values and store them in dedicated keyword fields, then reindex the data.

TL;DR: seems that for efficient aggregation we will need to modify the indexing, since what WikibaseCirrusSearch provides is not efficient for it. 

This works, but there has to be a better way:

```json
{
  "size": 0,
  "query": {
    "match": {
      "statement_keywords": "P1460=Q5976449"
    }
  },
  "aggs": {
    "volumes": {
      "terms": {
        "script": {
          "source": "def volumes = []; if (params['_source'].statement_keywords != null) { for (def statement : params['_source'].statement_keywords) { if (statement.startsWith('P16=')) { volumes.add(statement.substring(4)); } } } return volumes;",
          "lang": "painless"
        },
        "size": 10,
        "order": { "_count": "desc" }
      }
    }
  }
}
```

Output:

```json
{
  "took" : 19,
  "timed_out" : false,
  "_shards" : {
    "total" : 3,
    "successful" : 3,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 50,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "volumes" : {
      "doc_count_error_upper_bound" : 1,
      "sum_other_doc_count" : 135,
      "buckets" : [
        {
          "key" : "Q57520",
          "doc_count" : 7
        },
        {
          "key" : "Q57535",
          "doc_count" : 6
        },
        {
          "key" : "Q57432",
          "doc_count" : 4
        },
        {
          "key" : "Q57542",
          "doc_count" : 4
        },
        {
          "key" : "Q85858",
          "doc_count" : 4
        },
        {
          "key" : "Q72080",
          "doc_count" : 3
        },
        {
          "key" : "Q128712",
          "doc_count" : 2
        },
        {
          "key" : "Q128714",
          "doc_count" : 2
        },
        {
          "key" : "Q57521",
          "doc_count" : 2
        },
        {
          "key" : "Q57523",
          "doc_count" : 2
        }
      ]
    }
  }
}
```

**Comment by malberts** [comment]
For range queries on the current `statement_values`, ChatGPT has a warning about using queries like this:
```
{
  "query": {
    "script_score": {
      "query": { "match_all": {} },
      "script": {
        "source": """
        def result = 0;
        for (keyword in doc['statement_keywords']) {
          if (keyword.startsWith('P1=')) {
            def value = Integer.parseInt(keyword.substring(3));
            if (value > 1000 && value < 2000) {
              result = 1;
              break;
            }
          }
        }
        return result;
        """
      }
    }
  }
}
```

> This works without changing the schema but is computationally expensive and should only be used for small-scale queries.

**Comment by JeroenDeDauw** [comment]
Yes, o1 was also not happy with it when it came up with this approach. Belows output refers to the quoted text in the first comment in this PR.

![image](https://github.com/user-attachments/assets/f408e23b-df51-43e7-8790-13313d65a7fb)

**Comment by malberts** [comment]
A quick ChatGPT example of a runtime field extracting a value out of the `statement_keywords` list:
```
{
  "runtime_mappings": {
    "P1_value": {
      "type": "long",
      "script": {
        "source": """
        for (def keyword : doc['statement_keywords']) {
          if (keyword.startsWith('P1=')) {
            def value = keyword.substring(3); // Extract the part after 'P1='
            emit(Long.parseLong(value)); // Convert to number
          }
        }
        """
      }
    }
  },
  "query": {
    "range": {
      "P1_value": {
        "gte": 1000,
        "lte": 2000
      }
    }
  }
}
```

Which still involves some kind of looping. Not sure about the performance comparison of this verus the script.

**Comment by JeroenDeDauw** [comment]
I added my original query to the first comment

**Comment by malberts** [comment]
I should have been clearer. My queries here relate more specifically to range facet filtering with either "script" or "runtime_mappings", but with the point that both are likely suboptimal for involving some scripting, loops specifically. I know this is a separate issue, but ultimately to solve that it might require approach (3) in any case, which might affect what we choose to do for this issue.

**Comment by JeroenDeDauw** [comment]
Done by switching to our own indexes
